File: java/http-endpoint-service/src/main/java/HttpEndpointService.java
import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.http.javadsl.ConnectHttp;
import akka.http.javadsl.Http;
import akka.http.javadsl.model.HttpRequest;
import akka.http.javadsl.model.HttpResponse;
import akka.stream.ActorMaterializer;
import akka.stream.javadsl.Flow;
import api.HttpEndpointServiceApi;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;

public class HttpEndpointService {
    private static final int PORT = 8080;
    private static final String BOOTSTRAP_SERVERS = "localhost:9092";
    private static final String TOPIC = "jfall-gps-updates";

    public static void main(String args[]) {

        System.out.println("Starting the HTTP endpoint service");
        ActorSystem system = ActorSystem.create("routes");
        final ActorMaterializer materializer = ActorMaterializer.create(system);
        final Http http = Http.get(system);

        // create the producer
        Properties producerProperties = new Properties();
        producerProperties.put("bootstrap.servers", BOOTSTRAP_SERVERS);
        producerProperties.put("key.serializer", StringSerializer.class);
        producerProperties.put("value.serializer", StringSerializer.class);
        KafkaProducer<String, String> producer = new KafkaProducer<>(producerProperties);

        HttpEndpointServiceApi api = new HttpEndpointServiceApi(producer, TOPIC);

        final Flow<HttpRequest, HttpResponse, NotUsed> routeFlow = api.getRoute().flow(system, materializer);

        System.out.println("Start endpoint on port " + PORT);
        http.bindAndHandle(routeFlow, ConnectHttp.toHost("localhost", PORT), materializer);
    }
}

File: http-endpoint-service/src/main/scala/http_endpoint_service/HttpEndpointService.scala
package http_endpoint_service

import java.util.Properties

import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.stream.{ActorMaterializer, Materializer}
import http_endpoint_service.api.HttpEndpointServiceApi
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.common.serialization.StringSerializer

import scala.concurrent.ExecutionContext

/**
  * The Main of this application. Create endpoints to receive gps updates, setup a connection with Kafka and put
  * the gps updates there.
  */
object HttpEndpointService extends App {

  val port = 8080
  val bootstrapServers = scala.util.Properties.envOrElse("BOOTSTRAP_URL", "localhost:9092")
  val topic = scala.util.Properties.envOrElse("TOPIC", "jfall-gps-updates")

  println("Starting the HTTP endpoint service")

  // Some setup to get Akka HTTP up and running
  implicit val system: ActorSystem = ActorSystem()
  implicit val executionContext: ExecutionContext = system.dispatcher
  implicit val mat: Materializer = ActorMaterializer()

  // create the producer
  val producerProperties = new Properties()
  producerProperties.put("bootstrap.servers", bootstrapServers)
  producerProperties.put("key.serializer", classOf[StringSerializer].getName)
  producerProperties.put("value.serializer", classOf[StringSerializer].getName)
  val producer = new KafkaProducer[String, String](producerProperties)

  startServer()

  /**
    * Create endpoints and make them available so you can retrieve gps updates.
    */
  def startServer()= {
    println(s"Start endpoint on port $port")
    val api = new HttpEndpointServiceApi(producer, topic)
    Http().bindAndHandle(
      handler = api.route,
      interface = "0.0.0.0",
      port = port
    )
  }

}
